### DISCUSSION TITLE
typing: how to do mixins for multi inheritance?

### DISCUSSION THREAD

**Discussion by ewerybody** [microsoft/pylance-release#1955] [Q&A]
Hello there! 👋 sorry. me again.
So I got along pretty nicely now with all things typing and feel much more comfortable with it by the time ..
Well: Now I look at some mixin classes for multi inheritance. And I'm kinda stumped:

```py
class Base:
    def __init__(self) -> None:
        print(f'{self.__class__.__name__} initialized!')

    def fancy_function(self):
        print('"fancy_function" called!')

    def some_basic_func(self):
        print('"some_basic_func" called!')


class _Minxin:
    def some_basic_func(self):
        self.fancy_function()  # <-- Cannot access member .. 
        print('advanced "some_basic_func" called!')


class Fancy(_Minxin, Base):
    def __init__(self) -> None:
        super().__init__()


if __name__ == '__main__':
    b = Base()
    b.some_basic_func()
    f = Fancy()
    f.some_basic_func()
```
This works very nicely
```
Base initialized!
"some_basic_func" called!
Fancy initialized!
"fancy_function" called!
advanced "some_basic_func" called!   
```

But **OF COURSE** this says `Cannot access member "fancy_function" for type "_Mixin" Member "fancy_function" is unknown`
Yeah ... well. This is **supposed** to not be there! 
How would I annotate this most conveniently 🤷‍♀️

**Comment by erictraut** [comment]
A mix-in class should not assume anything about the classes it is combined with. Your mix-in implementation assumes that it is combined with another class that has a method called `fancy_function`, which is a fragile assumption. So my first recommendation is to avoid this pattern in your code.

If you don't want to switch to a less-fragile pattern, you can make the above code type check by telling the type checker about your assumption. To do this, you would define a protocol class, like this:

```python
class _SupportsFancyFunction(Protocol):
    def fancy_function(self):
        raise NotImplementedError()

class _Minxin(_SupportsFancyFunction):
    ...
```
